### PR TITLE
Fix for games with altered file extensions

### PR DESCRIPTION
**Pull request by sorlok**
Ready to merge (from my point of view). 

This is my implementation for Issue #1920; please see the initial discussion there.

Regarding @Ghabry's comments:
1) Check only 2 files: Done, but see note
2) Assume larger==LDB: Done
3) LcfXXX checks: Done, but see note

Some notes on this fix:
1) What's the easiest way to cheaply check if a directory *might* be a game project? Right now I check for RPG_RT.ini, but as you said that's not always present. I expect RPG_RT.exe might also be stripped. Is there any good file to check, or should I just disable this minor performance hack?  ---Side note: it might be worth keeping if *all games with hacked extensions* have .ini files.
2) The Meta file now has "LmtHeaderString=XXX", which allows us to pass the "XXX" string in to liblcf (i.e., to tell it what string to expect within the file header). As you said, errors are disabled if this string doesn't match  --but it still shows a warning. Do we want to allow passing in the expected header string to liblcf when parsing to silence the warning, or is this not worth the trouble?
3) We cannot apply the Meta file until *after* scanning the directory. Main reason: the Meta instance is stored in player.cpp. Sub reason: we don't know which file to calculate the CRC32 of until we scan (since they were renamed). This is not a huge problem, since the scanning is pretty fast.
4) I scan the names of all mapXXXX.ext files to ensure the extension is consistent across files. This might be overkill; we could (a) just stop on the first file or (b) check for the starting map filename (pulled from LMT), assuming we've parsed LMT at this point (which I think we have).
5) This fix assumes all "rpg_rt.XXX" files are renamed by extension ONLY --i.e., if you rename it to "abc.lmt" then this fix won't find it. Do any hacks do this, and do we care?
6) Just reiterating that the meta file can be used to overcome most limitations as discussed (e.g., slow lmu scanning and small LDB files).

### CODE REVIEW

**Comment by Ghabry** [comment]
One quick comment: std::regex is broken on some platforms and can't be used :(

**Comment by fdelapena** [comment]
Jenkins bot: ok to test

**Comment by fdelapena** [comment]
Jenkins: test this please

**Comment by fdelapena** [comment]
jenkins: test this please

**Comment by Ghabry** [comment]
Thanks for your contribution. Really cool to have support for this. Reviewing can currently take longer because it is summer and vacation time, so don't worry if there is less feedback . We are more active during the cold months ;).

**Comment by Ghabry** [comment]
(I was mentioning std::regex before but this should work properly since gcc5.0 so can be ignored)

- What does "Snh" mean?

> What's the easiest way to cheaply check if a directory might be a game project? Right now I check for RPG_RT.ini, but as you said that's not always present. I expect RPG_RT.exe might also be stripped.

When a game tree is parsed the file list is already in memory so the expensive operation (Read directory and alloc all the strings) is already done. I don't think doing some string testing will be really noticable. Also your code will only execute for the "not a game" case which shouldn't happen too often so don't optimize too much.

Sometimes games rename both RPG_RT and the ini. When the Exe is renamed it also changes the ini file it searches for (Game.exe will search for Game.ini)

> [LCF Header] As you said, errors are disabled if this string doesn't match --but it still shows a warning. Do we want to allow passing in the expected header string to liblcf when parsing to silence the warning, or is this not worth the trouble?

The "warning" is only printed on terminal so I don't consider this important

> I scan the names of all mapXXXX.ext files to ensure the extension is consistent across files.

I agree on you this is overkill. Just using the first mapfile you can find is good enough, we can still adopt if any game ever breaks this heuristic... (because there is some Map0001.bak there)

> [RPG_RT prefixed changed] Do any hacks do this, and do we care?

Not sure if any game does this. Is also a bit hard to analyse. When you provide me with some script I can use to scan hundreds of games easily I can tell you quickly ;)

**Comment by sorlok** [comment]
> What does "Snh" mean?
Just my initials; I usually prefix branches with this to minimize the chance of conflict. Let me know if you prefer an alternative naming scheme (I can file a new PR).

> I don't think doing some string testing will be really noticeable.
Great, I removed the INI check.

> The "warning" is only printed on terminal so I don't consider this important
Great, I removed this from the sample .ini file

> Just using the first mapfile you can find is good enough
Done

> When you provide me with some script I can use to scan hundreds of games easily I can tell you quickly
That's not a bad idea. I'll write a quick Python script that can check this stuff.

**Comment by sorlok** [comment]
Rebased and squashed. This branch is ready for merge (from my point of view --let me know if anything needs to change).

**Comment by sorlok** [comment]
Hey all,

Just got back from vacation; wanted to check if there's anything more needed with this branch. (Also if you are all still on vacation that's fine too.) Thanks!

**Comment by Ghabry** [comment]
I'm back from vacation and currently prepare running your script (extracting all the games takes a while \*g\*)

**Comment by Ghabry** [comment]
Some games are molebox encrypted or have corrupted zip paths (\ instead of /) but for ones that work (around 1900): Only 2 are affected by this PR:

- Beyond Black Horizon (https://rmarchiv.de/games/443) - This has changed file extensions
- Verloren (https://rmarchiv.de/games/1333) - Uses default extensions but the maps are in a folder "MM" with names 0001.lmu etc. _out of scope_

**Comment by sorlok** [comment]
> I'm back from vacation and currently prepare running your script (extracting all the games takes a while *g*)
> 
> Minor patch (recurses in a subdirectory when the subdir only has one folder)
> 
> ```
> 91,96c91
> <   dirlist = list(os.listdir(full_dir))
> <   dirs = list(filter(lambda x: os.path.isdir(os.path.join(full_dir, x)), dirlist))
> <   if len(list(dirs)) == 1:
> <     return scan_project(full_dir, dirs[0])
> < 
> <   for filepath in dirlist:
> ---
> >   for filepath in os.listdir(full_dir):
> ```

Updated, thanks

**Comment by Ghabry** [comment]
This needs now a minor rebase

File: src/filefinder.cpp
Comment (Ghabry) [inline src/filefinder.cpp:760]: this doesn't work when launched from the Gamebrowser (wrong relative path).

Use:
```
candidates.push_back({item.second, ext, GetFileSize(FileFinder::FindDefault(dir, item.second))});
```

**Comment by Ghabry** [comment]
In my first quick test it worked. Good job :)

File: src/filefinder.cpp
Comment (sorlok) [reply src/filefinder.cpp]: Fixed in 9e361a9bf1052, thanks

**Comment by sorlok** [comment]
Rebased; I'm running a few quick tests (and waiting on builds).

**Comment by sorlok** [comment]
Tested again on Linux and Android; confirmed it's working.

File: src/filefinder.cpp
Comment (Ghabry) [inline src/filefinder.cpp:758]: Can you move the entire guess/remap logic in a new file? My problem here is that I have a huge PR touching FileFinder in the works and such a huge change is terrible to handle for me :(.

No idea. Make a name up: "filefinder_guesser.cpp/h", "file_guesser.cpp/h", "project_guesser.cpp/.h".

Move ``GetRPG2kProjectWithRenames, GuessAndAddLmuExtension`` and ``RPG2KNonStandardFilenameGuesser, RPG2KFileExtRemap`` to this

``IsRPG2kProjectWithRenames`` can stay as it fits with the rest.

File: builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameScanner.java
Comment (sorlok) [reply builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameScanner.java]: Done.

File: builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameBrowserHelper.java
Comment (sorlok) [reply builds/android/app/src/main/java/org/easyrpg/player/game_browser/GameBrowserHelper.java]: Done.

File: src/game_map.cpp
Comment (sorlok) [reply src/game_map.cpp]: Done.

File: src/game_map.cpp
Comment (sorlok) [reply src/game_map.cpp]: Done, thanks for the info (I don't know much about this output api).

File: src/game_map.cpp
Comment (sorlok) [reply src/game_map.cpp]: Done.

File: src/options.h
Comment (sorlok) [reply src/options.h]: Fixed. Sorry about that, it was not intentional.

File: src/filefinder.cpp
Comment (sorlok) [reply src/filefinder.cpp]: Done.

**Comment by sorlok** [comment]
Changes made and rebased. Tested and works on Linux. I'm still building on Android.

**Comment by sorlok** [comment]
Android is working great. That's everything on my side.

File: src/filext_guesser.h
Comment (Ghabry) [inline src/filext_guesser.h:97]: new line at end of file missing

File: src/filext_guesser.h
Comment (Ghabry) [inline src/filext_guesser.h:32]: I think there is a typo here. Should be FileExtGuesser and the file fileext_guesser.

**Review by Ghabry — COMMENTED** [review]
Except for my remaining nits this looks now fine for me.

Do you know how to "squash" in a rebase? If yes please squash the two "Code review" commits into one.

I marked it as "Has PR deps" because I want to merge this after #2293
(that PR changes some string functions and want to attempt a regex -> starts/ends_with change as I expect this to be faster here)

File: src/filext_guesser.h
Comment (sorlok) [reply src/filext_guesser.h]: Fixed.

File: src/filext_guesser.h
Comment (sorlok) [reply src/filext_guesser.h:32]: Hmm, I was thinking "filext", but I now think that might be trademarked by the associated ".com" address. I'll change this after work.

**Comment by sorlok** [comment]
Renamed filext to fileext (and similar).
Switched from regex to string_view.
Rebased, squashed.

Code seems like it's still working (tested loading two games on Linux).

File: src/filefinder.cpp
Comment (Ghabry) [inline src/filefinder.cpp:29]: unnecessary include (thanks for changing this :+1: )

**Review by Ghabry — COMMENTED** [review]
Want to quickly test this but it looks now fine for me! Thanks! Very useful feature :)

The commit message can be improved. Instead of "Squashed commits" can you use something like "Support games with altered file extensions" (you can keep the verbose message part as is)

File: src/filefinder.cpp
Comment (sorlok) [reply src/filefinder.cpp]: Fixed (and you're welcome!)

Comment (sorlok) [reply src/filefinder.cpp]: Also fixed the commit message.

**Review by Ghabry — CHANGES_REQUESTED** [review]
I planned to merge it but I just found an important bug (sorry).

When starting a renamed game the output is "Debug: Encoding not detected". This means that for non-English games the language detection is broken (and they won't run). The encoding detection still only uses RPG_RT.ldb.

I think you must reorder the code a bit. The following must happen before "GetEncoding" is called:

1. Load Meta file
2. Guess extensions
3. GetEncoding
4. [...] stuff
5. LoadDatabase

**Comment by sorlok** [comment]
Ah, interesting. Glad you caught this bug before merging! I'm a bit busy this weekend, but I'll check it out next week.

**Comment by Ghabry** [comment]
About the build failures: That was a very bad timing for updating the PR. There is currently a breaking change from liblcf pending/rebuilding.
You will have to rebase again in ~1h ^^'.

**Comment by sorlok** [comment]
> About the build failures: That was a very bad timing for updating the PR. There is currently a breaking change from liblcf pending/rebuilding.
> You will have to rebase again in ~1h ^^'.

No worries; I'm still working on the fix.

**Comment by sorlok** [comment]
Ok, the fix is in (tested on Laxius Power and a standard game).
1) I left it as its own commit for now, since I think that makes it easier to review. Let me know if/when you want it squashed.
2) Are there any EasyRPG format games I can test this on? I want to make sure I didn't break anything with the new format.

**Review by Ghabry — APPROVED** [review]
This works perfect now. Tested it. Thanks for this great contribution!

**Comment by Ghabry** [comment]
@sorlok 
FYI
Made an announcement: https://twitter.com/EasyRPG/status/1301108541185548288

**Comment by sorlok** [comment]
Cool, thanks!

**Pull request merged**
